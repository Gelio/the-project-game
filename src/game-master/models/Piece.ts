import { Point } from '../../common/Point';

export class Piece {
  public isSham: boolean;

  /**
   * When the piece is picked up, this is the position of the player
   */
  public position: Point;
  public isPickedUp: boolean;
}
