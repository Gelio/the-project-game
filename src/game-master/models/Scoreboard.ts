export class Scoreboard {
  public team1Score: number;
  public team2Score: number;
  public readonly scoreLimit: number;

  constructor(scoreLimit: number) {
    this.scoreLimit = scoreLimit;
    this.team1Score = 0;
    this.team2Score = 0;
  }
}
