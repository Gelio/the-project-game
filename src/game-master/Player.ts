import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
