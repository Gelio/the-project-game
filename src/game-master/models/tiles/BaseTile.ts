import { Player } from '../../Player';
import { Piece } from '../Piece';

export abstract class BaseTile {
  public readonly type: string;

  public x: number;
  public y: number;

  public player: Player | null = null;
  public piece: Piece | null = null;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}
