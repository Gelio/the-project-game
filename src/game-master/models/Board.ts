import { Point } from '../../common/Point';
import { BoardSize } from '../../interfaces/BoardSize';
import { Player } from '../Player';
import { Piece } from './Piece';
import { Tile } from './tiles/Tile';

export class Board {
  public readonly size: BoardSize;
  public readonly tiles: Tile[][];
  public readonly pieces: Piece[] = [];

  constructor(size: BoardSize, tiles: Tile[][]) {
    this.size = size;
    this.tiles = tiles;
  }

  public reset() {
    // TODO: implement this
  }

  public getTileAtPosition(position: Point) {
    const tileColumn = this.tiles[position.x];
    if (!tileColumn) {
      throw new Error('Invalid X coordinate');
    }

    const tile = tileColumn[position.y];
    if (!tile) {
      throw new Error('Invalid Y coordinate');
    }

    return tile;
  }

  public addPlayer(player: Player) {
    this.getTileAtPosition(player.position).player = player;
  }

  public removePlayer(point: Point) {
    this.tiles[point.x][point.y].player = null;
  }

  public movePlayer(player: Player, newPosition: Point) {
    const previousTile = this.getTileAtPosition(player.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.player !== player) {
      throw new Error('Old player position corrupted');
    }

    if (newTile.player) {
      throw new Error('Two players cannot stand on the same tile');
    }

    previousTile.player = null;
    newTile.player = player;
  }

  public addPiece(piece: Piece) {
    if (this.pieces.indexOf(piece) !== -1) {
      throw new Error('Piece already added');
    }

    const tile = this.getTileAtPosition(piece.position);
    if (tile.piece) {
      throw new Error('Piece already exists at that position');
    }

    tile.piece = piece;
    this.pieces.push(piece);
  }

  public removePiece(piece: Piece) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    this.pieces.splice(index, 1);

    if (!piece.isPickedUp) {
      const tile = this.getTileAtPosition(piece.position);
      tile.piece = null;
    }
  }

  public movePiece(piece: Piece, newPosition: Point) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    const previousTile = this.getTileAtPosition(piece.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.piece !== piece) {
      throw new Error('Old piece position corrupted');
    }

    if (newTile.piece) {
      throw new Error('Cannot move a piece on a tile which already has one');
    }

    previousTile.piece = null;
    newTile.piece = piece;
    piece.position = newPosition;
  }
}
