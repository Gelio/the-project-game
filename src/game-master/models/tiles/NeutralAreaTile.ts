import { BaseTile } from './BaseTile';

export class NeutralAreaTile extends BaseTile {}
