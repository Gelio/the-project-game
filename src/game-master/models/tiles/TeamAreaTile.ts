import { BaseTile } from './BaseTile';

export class TeamAreaTile extends BaseTile {
  public readonly type = 'TeamAreaTile';
  public hasGoal = false;
  public hasCompletedGoal = false;
}
