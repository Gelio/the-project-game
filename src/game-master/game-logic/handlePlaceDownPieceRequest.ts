import { createDelay } from '../../common/createDelay';
import { GAME_MASTER_ID } from '../../common/EntityIds';
import { Point } from '../../common/Point';

import { PlaceDownPieceRequest } from '../../interfaces/requests/PlaceDownPieceRequest';
import { PlaceDownPieceResponse } from '../../interfaces/responses/PlaceDownPieceResponse';

import { Player } from '../Player';
import { ProcessMessageResult } from '../ProcessMessageResult';

import { Piece } from '../models/Piece';
import { NeutralAreaTile } from '../models/tiles/NeutralAreaTile';
import { TeamAreaTile } from '../models/tiles/TeamAreaTile';
import { MessageHandlerDependencies } from './MessageHandlerDependencies';

function createPlaceDownPieceResponse(
  recipientId: string,
  didCompleteGoal?: boolean
): PlaceDownPieceResponse {
  return {
    type: 'PLACE_DOWN_PIECE_RESPONSE',
    payload: {
      didCompleteGoal
    },
    recipientId: recipientId,
    senderId: GAME_MASTER_ID
  };
}

function handlePlaceDownOnNeutralAreaTile(
  tile: NeutralAreaTile,
  player: Player,
  piece: Piece,
  playerPosition: Point,
  { actionDelays }: MessageHandlerDependencies
): ProcessMessageResult<PlaceDownPieceResponse> {
  tile.piece = piece;
  piece.isPickedUp = false;
  piece.position = playerPosition.clone();
  player.heldPiece = null;

  return {
    valid: true,
    delay: actionDelays.place,
    responseMessage: createDelay(actionDelays.place).then(() =>
      createPlaceDownPieceResponse(player.playerId, undefined)
    )
  };
}

function handlePlaceDownOnTeamAreaTile(
  tile: TeamAreaTile,
  player: Player,
  piece: Piece,
  playerPosition: Point,
  { actionDelays, board, scoreboard, logger, onPointsLimitReached }: MessageHandlerDependencies
): ProcessMessageResult<PlaceDownPieceResponse> {
  board.removePiece(piece);
  player.heldPiece = null;
  piece.isPickedUp = false;

  if (piece.isSham) {
    logger.verbose(`Player ${player.playerId} placed down a sham at ${playerPosition.toString()}`);

    return {
      valid: true,
      delay: actionDelays.place,
      responseMessage: createDelay(actionDelays.place).then(() =>
        createPlaceDownPieceResponse(player.playerId, undefined)
      )
    };
  }

  if (!tile.hasGoal) {
    logger.verbose(`Player ${player.playerId} placed down a piece at ${playerPosition.toString()}`);

    return {
      valid: true,
      delay: actionDelays.place,
      responseMessage: createDelay(actionDelays.place).then(() =>
        createPlaceDownPieceResponse(player.playerId, false)
      )
    };
  }

  if (!tile.hasCompletedGoal) {
    logger.verbose(`Player ${player.playerId} completed a goal at ${playerPosition.toString()}`);
    tile.hasCompletedGoal = true;

    if (player.teamId === 1) {
      scoreboard.team1Score++;
      if (scoreboard.team1Score === scoreboard.scoreLimit) {
        onPointsLimitReached();
      }
    } else if (player.teamId === 2) {
      scoreboard.team2Score++;
      if (scoreboard.team2Score === scoreboard.scoreLimit) {
        onPointsLimitReached();
      }
    }

    return {
      valid: true,
      delay: actionDelays.place,
      responseMessage: createDelay(actionDelays.place).then(() =>
        createPlaceDownPieceResponse(player.playerId, true)
      )
    };
  }

  logger.verbose(
    `Player ${player.playerId} placed down a piece at ${playerPosition.toString()} (completed goal)`
  );

  return {
    valid: true,
    delay: actionDelays.place,
    responseMessage: createDelay(actionDelays.place).then(() =>
      createPlaceDownPieceResponse(player.playerId, false)
    )
  };
}

export function handlePlaceDownPieceRequest(
  dependencies: MessageHandlerDependencies,
  sender: Player,
  _discoveryRequest: PlaceDownPieceRequest
): ProcessMessageResult<PlaceDownPieceResponse> {
  const { board, logger } = dependencies;
  const { position } = sender;

  if (!position) {
    logger.error('Player position not set');

    return {
      valid: false,
      reason: 'Internal GM error. Invalid player position'
    };
  }

  if (!sender.heldPiece) {
    return {
      valid: false,
      reason: 'Player does not hold a piece'
    };
  }

  const tile = board.getTileAtPosition(position);
  if (tile.piece) {
    return {
      valid: false,
      reason: 'Tile under the player already has a piece'
    };
  }

  if (tile.type === 'NeutralAreaTile') {
    return handlePlaceDownOnNeutralAreaTile(tile, sender, sender.heldPiece, position, dependencies);
  }

  if (tile.type === 'TeamAreaTile') {
    return handlePlaceDownOnTeamAreaTile(tile, sender, sender.heldPiece, position, dependencies);
  }

  logger.error(`Invalid tile type '${(<any>tile).type}' at ${position.toString()}`);

  return {
    valid: false,
    reason: 'Internal GM error. Invalid tile type'
  };
}
