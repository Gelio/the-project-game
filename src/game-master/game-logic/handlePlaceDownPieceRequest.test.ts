import { LoggerInstance } from 'winston';

import { LoggerFactory } from '../../common/logging/LoggerFactory';
import { Point } from '../../common/Point';

import { ActionDelays } from '../../interfaces/ActionDelays';
import { PlaceDownPieceResponse } from '../../interfaces/responses/PlaceDownPieceResponse';

import { Board } from '../models/Board';
import { Piece } from '../models/Piece';
import { Scoreboard } from '../models/Scoreboard';
import { TeamAreaTile } from '../models/tiles/TeamAreaTile';

import { Player } from '../Player';
import {
  InvalidMessageResult,
  ProcessMessageResult,
  ValidMessageResult
} from '../ProcessMessageResult';

import { handlePlaceDownPieceRequest } from './handlePlaceDownPieceRequest';

function valid(
  result: ProcessMessageResult<PlaceDownPieceResponse>
): ValidMessageResult<PlaceDownPieceResponse> {
  return <any>result;
}

function invalid(result: ProcessMessageResult<PlaceDownPieceResponse>): InvalidMessageResult {
  return <any>result;
}

describe('[GM] handlePlaceDownPiece', () => {
  let board: Board;
  let actionDelays: ActionDelays;
  let player: Player;
  let piece: Piece;
  let scoreboard: Scoreboard;
  let logger: LoggerInstance;

  beforeEach(() => {
    board = new Board(
      {
        goalArea: 10,
        taskArea: 20,
        x: 10
      },
      10
    );

    actionDelays = <any>{
      place: 200
    };

    player = new Player();
    player.isBusy = true;
    player.playerId = 'player1';
    player.position = new Point(0, 0);

    piece = new Piece();
    piece.isPickedUp = true;
    piece.isSham = false;
    piece.position = player.position.clone();
    player.heldPiece = piece;
    board.addPiece(piece);

    const loggerFactory = new LoggerFactory();
    loggerFactory.logLevel = 'error';

    scoreboard = new Scoreboard(5);

    logger = loggerFactory.createEmptyLogger();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  function executeHandler(): ProcessMessageResult<PlaceDownPieceResponse> {
    return handlePlaceDownPieceRequest(
      {
        board,
        playersContainer: <any>{},
        actionDelays,
        logger,
        scoreboard,
        sendMessage: jest.fn()
      },
      player,
      <any>{}
    );
  }

  it('should mark action as invalid when player does not hold a piece', () => {
    player.heldPiece = null;
    const result = executeHandler();

    expect(result.valid).toBe(false);
    expect(invalid(result).reason).toMatchSnapshot();
  });

  it('should mark action as invalid when tile already has a piece', () => {
    board.getTileAtPosition(<Point>player.position).piece = new Piece();
    const result = executeHandler();

    expect(result.valid).toBe(false);
    expect(invalid(result).reason).toMatchSnapshot();
  });

  describe('on neutral tile', () => {
    beforeEach(() => {
      player.position = new Point(0, board.size.goalArea);
      piece.position = player.position.clone();
    });

    it('should place the piece on the board', () => {
      executeHandler();
      const tile = board.getTileAtPosition(<Point>player.position);

      expect(tile.piece).toBe(piece);
      expect(piece.isPickedUp).toBe(false);
    });

    it('should take the piece from the player', () => {
      executeHandler();

      expect(player.heldPiece).toBe(null);
    });

    it('should be valid', () => {
      const result = executeHandler();

      expect(result.valid).toBe(true);
    });

    it('should have a correct delay', () => {
      const result = executeHandler();

      expect(valid(result).delay).toBe(actionDelays.place);
    });

    it('should resolve the message after a delay', () => {
      const result = executeHandler();

      valid(result).responseMessage.then(response => {
        expect(response.recipientId).toBe(player.playerId);
      });

      expect.assertions(1);
      jest.advanceTimersByTime(actionDelays.place);
    });

    it('should mark didCompleteGoal as undefined', async () => {
      const result = executeHandler();

      jest.advanceTimersByTime(actionDelays.place);

      const response = await valid(result).responseMessage;
      expect(response.payload.didCompleteGoal).toBeUndefined();
    });

    it('should not resolve the message before the delay', () => {
      const result = executeHandler();
      let resolved = false;

      valid(result).responseMessage.then(() => {
        resolved = true;
      });
      jest.advanceTimersByTime(actionDelays.place - 1);

      expect(resolved).toBe(false);
    });

    it('should not modify team scores', () => {
      executeHandler();

      expect(scoreboard.team1Score).toBe(0);
      expect(scoreboard.team2Score).toBe(0);
    });
  });

  describe('on team area tile', () => {
    let tile: TeamAreaTile;

    beforeEach(() => {
      tile = <TeamAreaTile>board.getTileAtPosition(<Point>player.position);
    });

    it('should remove the piece from the board', () => {
      executeHandler();

      expect(board.pieces).not.toContain(piece);
    });

    it('should take the piece from the player', () => {
      executeHandler();

      expect(player.heldPiece).toBe(null);
    });

    it('should mark piece as not picked up', () => {
      executeHandler();

      expect(piece.isPickedUp).toBe(false);
    });

    it('should be valid', () => {
      const result = executeHandler();

      expect(result.valid).toBe(true);
    });

    it('should have a correct delay', () => {
      const result = executeHandler();

      expect(valid(result).delay).toBe(actionDelays.place);
    });

    it('should resolve the message after a delay', async () => {
      const result = executeHandler();
      jest.advanceTimersByTime(actionDelays.place);

      const response = await valid(result).responseMessage;
      expect(response.recipientId).toBe(player.playerId);
    });

    it('should not resolve the message before the delay', () => {
      const result = executeHandler();
      let resolved = false;

      valid(result).responseMessage.then(() => {
        resolved = true;
      });
      jest.advanceTimersByTime(actionDelays.place - 1);

      expect(resolved).toBe(false);
    });

    describe('when the piece is a sham', () => {
      beforeEach(() => {
        piece.isSham = true;
      });

      it('should mark didCompleteGoal as undefined', async () => {
        const result = executeHandler();

        jest.advanceTimersByTime(actionDelays.place);

        const response = await valid(result).responseMessage;
        expect(response.payload.didCompleteGoal).toBeUndefined();
      });
    });

    describe('when the piece is a valid piece', () => {
      describe('and tile does not have a goal', () => {
        beforeEach(() => {
          tile.hasGoal = false;
        });

        it('should mark didCompleteGoal as false', async () => {
          const result = executeHandler();

          jest.advanceTimersByTime(actionDelays.place);

          const response = await valid(result).responseMessage;
          expect(response.payload.didCompleteGoal).toBe(false);
        });
      });

      describe('and tile does have an uncompleted goal', () => {
        beforeEach(() => {
          tile.hasGoal = true;
          tile.hasCompletedGoal = false;
        });

        it('should mark didCompleteGoal as true', async () => {
          const result = executeHandler();

          jest.advanceTimersByTime(actionDelays.place);

          const response = await valid(result).responseMessage;
          expect(response.payload.didCompleteGoal).toBe(true);
        });

        it('should increase the score for team 1', () => {
          player.teamId = 1;
          executeHandler();

          expect(scoreboard.team1Score).toBe(1);
        });

        it("should not modify the other team's score", () => {
          player.teamId = 1;
          executeHandler();

          expect(scoreboard.team2Score).toBe(0);
        });

        it('should increase the score for team 2', () => {
          player.teamId = 2;
          executeHandler();

          expect(scoreboard.team2Score).toBe(1);
        });

        // TODO: add a test for a win condition
      });

      describe('and tile does have a completed goal', () => {
        beforeEach(() => {
          tile.hasGoal = true;
          tile.hasCompletedGoal = true;
        });

        it('should mark didCompleteGoal as false', async () => {
          const result = executeHandler();

          jest.advanceTimersByTime(actionDelays.place);

          const response = await valid(result).responseMessage;
          expect(response.payload.didCompleteGoal).toBe(false);
        });

        it("should not increase the team's score", () => {
          player.teamId = 1;
          executeHandler();

          expect(scoreboard.team1Score).toBe(0);
        });
      });
    });
  });
});
