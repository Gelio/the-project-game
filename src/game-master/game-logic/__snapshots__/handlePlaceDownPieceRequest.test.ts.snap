// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] handlePlaceDownPiece should mark action as invalid when player does not hold a piece 1`] = `"Player does not hold a piece"`;

exports[`[GM] handlePlaceDownPiece should mark action as invalid when tile already has a piece 1`] = `"Tile under the player already has a piece"`;
